packages/driver/src/cy/multi-domain/index.ts

```diff
@@ -135,51 +133,54 @@ export function addCommands (Commands, Cypress: Cypress.Cypress, cy: Cypress.cy,
-              state: {
-                viewportWidth: Cypress.state('viewportWidth'),
-                viewportHeight: Cypress.state('viewportHeight'),
-                runnable: serializeRunnable(Cypress.state('runnable')),
-                duringUserTestExecution: Cypress.state('duringUserTestExecution'),
-                hookId: state('hookId'),
-              },
+        communicator.once('bridge:ready', (_data, bridgeReadyDomain) => {
+          if (bridgeReadyDomain === domain) {
+            // now that the spec bridge is ready, instantiate Cypress with the current app config and environment variables for initial sync when creating the instance
+            communicator.toSpecBridge(domain, 'initialize:cypress', {
               config: preprocessConfig(Cypress.config()),
               env: preprocessEnv(Cypress.env()),
             })
-          } catch (err: any) {
-            const wrappedErr = $errUtils.errByPath('switchToDomain.run_domain_fn_errored', {
-              error: err.message,
-            })
 
-            cleanup()
-            reject(wrappedErr)
-          } finally {
-            // @ts-ignore
-            cy.isAnticipatingMultiDomain(false)
+            state('readyForMultiDomain', true)
+
+            // once the secondary domain page loads, send along the
+            // user-specified callback to run in that domain
+            try {
+              communicator.toSpecBridge(domain, 'run:domain:fn', {
+                data,
+                fn: callbackFn.toString(),
+                // let the spec bridge version of Cypress know if config read-only values can be overwritten since window.top cannot be accessed in cross-origin iframes
+                // this should only be used for internal testing. Cast to boolean to guarantee serialization
+                // @ts-ignore
+                skipConfigValidation: !!window.top.__cySkipValidateConfig,
+                state: {
+                  viewportWidth: Cypress.state('viewportWidth'),
+                  viewportHeight: Cypress.state('viewportHeight'),
+                  runnable: serializeRunnable(Cypress.state('runnable')),
+                  duringUserTestExecution: Cypress.state('duringUserTestExecution'),
+                  hookId: state('hookId'),
+                },
+                config: preprocessConfig(Cypress.config()),
+                env: preprocessEnv(Cypress.env()),
+              })
+            } catch (err: any) {
+              const wrappedErr = $errUtils.errByPath('switchToDomain.run_domain_fn_errored', {
+                error: err.message,
+              })
+
+              reject(wrappedErr)
+            } finally {
+              // @ts-ignore
+              cy.isAnticipatingMultiDomain(false)
+            }
           }
         })
 
         // this signals to the runner to create the spec bridge for
         // the specified domain
-        communicator.emit('expect:domain', domain)
+        Cypress.emit('expect:domain', domain)
```

Comment:
I changed this because it made more sense to me that cypress is sending the message than the multi-domain communicator

Reply:
No problem, I can change it back, though it still feels weird to use the communicator here.

Reply:
Reverted